horovod/common/ops/nccl_operations.cc

```diff
@@ -0,0 +1,293 @@
+// Copyright 2016 The TensorFlow Authors. All Rights Reserved.
+// Modifications copyright (C) 2019 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#include "nccl_operations.h"
+
+namespace horovod {
+namespace common {
+
+ncclDataType_t GetNCCLDataType(const std::shared_ptr<Tensor> tensor) {
+  switch (tensor->dtype()) {
+    case HOROVOD_INT32:
+      return ncclInt32;
+    case HOROVOD_INT64:
+      return ncclInt64;
+    case HOROVOD_FLOAT16:
+      return ncclFloat16;
+    case HOROVOD_FLOAT32:
+      return ncclFloat32;
+    case HOROVOD_FLOAT64:
+      return ncclFloat64;
+    default:
+      throw std::logic_error("Type " + DataType_Name(tensor->dtype()) +
+                             " is not supported in NCCL mode.");
+  }
+}
+
+void NCCLContext::ErrorCheck(std::string op_name, ncclResult_t nccl_result) {
+  if (nccl_result != ncclSuccess) {
+    throw std::logic_error(std::string(op_name) + " failed: " + ncclGetErrorString(nccl_result));
+  }
+}
+
+
+NCCLAllreduce::NCCLAllreduce(NCCLContext* nccl_context,
+                             CUDAContext* cuda_context,
+                             CommunicationContext* comm_context,
```

Comment:
`Orchestration`

Reply:
Changed to Channel to more accurately reflect its role as a separate collective communication channel, and made its use limited to specific frameworks (NCCL).